crates/ruff_linter/src/rules/flake8_pie/rules/unnecessary_dict_kwargs.rs

```diff
@@ -40,41 +41,83 @@ use crate::checkers::ast::Checker;
 #[violation]
 pub struct UnnecessaryDictKwargs;
 
-impl Violation for UnnecessaryDictKwargs {
+impl AlwaysFixableViolation for UnnecessaryDictKwargs {
     #[derive_message_formats]
     fn message(&self) -> String {
         format!("Unnecessary `dict` kwargs")
     }
+
+    fn fix_title(&self) -> String {
+        format!("Remove unnecessary kwargs")
+    }
 }
 
 /// PIE804
 pub(crate) fn unnecessary_dict_kwargs(checker: &mut Checker, expr: &Expr, kwargs: &[Keyword]) {
     for kw in kwargs {
         // keyword is a spread operator (indicated by None)
-        if kw.arg.is_none() {
-            if let Expr::Dict(ast::ExprDict { keys, .. }) = &kw.value {
-                // ensure foo(**{"bar-bar": 1}) doesn't error
-                if keys.iter().all(|expr| expr.as_ref().is_some_and( is_valid_kwarg_name)) ||
-                    // handle case of foo(**{**bar})
-                    (keys.len() == 1 && keys[0].is_none())
-                {
-                    let diagnostic = Diagnostic::new(UnnecessaryDictKwargs, expr.range());
-                    checker.diagnostics.push(diagnostic);
-                }
+        if kw.arg.is_some() {
+            continue;
+        }
+
+        let Expr::Dict(ast::ExprDict { keys, values, .. }) = &kw.value else {
+            continue;
+        };
+
+        // Ex) `foo(**{**bar})`
+        if matches!(keys.as_slice(), [None]) {
+            let mut diagnostic = Diagnostic::new(UnnecessaryDictKwargs, expr.range());
+
+            if checker.patch(diagnostic.kind.rule()) {
+                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                    format!("**{}", checker.locator().slice(values[0].range())),
+                    kw.range(),
+                )));
             }
+
+            checker.diagnostics.push(diagnostic);
+            continue;
         }
+
+        // Ensure that every keyword is a valid keyword argument (e.g., avoid errors for cases like
+        // `foo(**{"bar-bar": 1})`).
+        let kwargs = keys
+            .iter()
+            .filter_map(|key| key.as_ref().and_then(as_kwarg))
+            .collect::<Vec<_>>();
```

Comment:
I restructured it such that, when we validate the keyword arguments, we also return the valid keyword argument _name_. That lets us avoid a bunch of unwraps below. In other words: when we validate, we return the validated data.

Reply:
ooo. Love it! Thanks!